### ISSUE TITLE
Detecting only one class causes terrible problems.

### ISSUE THREAD

**Issue by kemingzeng** [pierluigiferrari/ssd_keras#154]
I implemented it on my own dataset, and this dataset has **only one class**(just vehicles), so I set n_classes to 1.
Not changed any hyper-parameters of **SSD512**, I got terrible consequences after 120 epochs iterations starting from the **VGG_ILSVRC_16_layers_fc_reduced.h5**.
you can see this one:
![](https://raw.githubusercontent.com/kemingzeng/kemingzeng.github.io/master/assets/image/Figure_2.png)
Do you have any ideas to help me?
I've checked:
1. input format is *'corners'*. (1069*500), here I use parse_csv

0 | 1 | 2 | 3 | 4 | 5
-- | -- | -- | -- | -- | --
3333bf6f-57cb-4d90-afe1-636dcd88a255.jpg | 1 | 498 | 237 | 671 | 409
3333bf6f-57cb-4d90-afe1-636dcd88a255.jpg | 1 | 1049 | 20 | 1067 | 102
1eb97c5b-78ef-455b-8c9c-3fc8c33201f9.jpg | 1 | 438 | 79 | 671 | 364
1eb97c5b-78ef-455b-8c9c-3fc8c33201f9.jpg | 1 | 627 | 52 | 741 | 153
1eb97c5b-78ef-455b-8c9c-3fc8c33201f9.jpg | 1 | 205 | 191 | 371 | 355
1eb97c5b-78ef-455b-8c9c-3fc8c33201f9.jpg | 1 | 822 | 0 | 962 | 41
1eb97c5b-78ef-455b-8c9c-3fc8c33201f9.jpg | 1 | 477 | 373 | 885 | 500
1eb97c5b-78ef-455b-8c9c-3fc8c33201f9.jpg | 1 | 1046 | 0 | 1067 | 32
eeeb4783-6acb-4e83-9adf-fc0d2ee1f74d.jpg | 1 | 195 | 266 | 371 | 409

2. Because I guess it maybe be caused by sample balance, so I tried to changed neg_pos_ratio=3 to neg_pos_ratio=1, but still get bad dectection.
![](https://raw.githubusercontent.com/kemingzeng/kemingzeng.github.io/master/assets/image/Figure_22.png)

Just want to die.  :sob:

**Comment by pierluigiferrari** [comment]
Looks a lot like a coordinate format issue. Did you set the `input_format` argument correctly in `parse_csv()`? In your case it should be `['image_name', 'class_id', 'xmin', 'ymin', 'xmax', 'ymax']`.

**Comment by kemingzeng** [comment]
@pierluigiferrari 
yes, I'm sure I used the right format:
```python
train_dataset.parse_csv(images_dir=train_dir, labels_filename=train_labels_csv,
                        input_format=['image_name', 'class_id', 'xmin', 'ymin', 'xmax', 'ymax'])
val_dataset.parse_csv(images_dir=val_dir, labels_filename=val_labels_csv,
                      input_format=['image_name', 'class_id', 'xmin', 'ymin', 'xmax', 'ymax'])
```

### Here is my complete codes:
```python
def train_main():

    # 1: Build the Keras model.

    K.clear_session() # Clear previous models from memory.

    model = ssd_512(image_size=(img_height, img_width, img_channels),
                    n_classes=n_classes,
                    mode='training',
                    l2_regularization=0.0005,
                    scales=scales,
                    aspect_ratios_per_layer=aspect_ratios,
                    two_boxes_for_ar1=two_boxes_for_ar1,
                    steps=steps,
                    offsets=offsets,
                    clip_boxes=clip_boxes,
                    variances=variances,
                    normalize_coords=normalize_coords,
                    subtract_mean=mean_color,
                    swap_channels=swap_channels)

    # 2: Load some weights into the model.

    # TODO: Set the path to the weights you want to load.
    weights_path = 'VGG_ILSVRC_16_layers_fc_reduced.h5'
    model.load_weights(weights_path, by_name=True)

    # 3: Instantiate an optimizer and the SSD loss function and compile the model.
    #    If you want to follow the original Caffe implementation, use the preset SGD
    #    optimizer, otherwise I'd recommend the commented-out Adam optimizer.

    sgd = SGD(lr=0.001, momentum=0.9, decay=0.0, nesterov=False)

    ssd_loss = SSDLoss(neg_pos_ratio=3, alpha=1.0)

    model.compile(optimizer=sgd, loss=ssd_loss.compute_loss)

    # data
    train_dataset = DataGenerator(load_images_into_memory=False, hdf5_dataset_path=None)
    val_dataset = DataGenerator(load_images_into_memory=False, hdf5_dataset_path=None)
    train_dir = 'train_1w'
    val_dir = 'train_b'
    train_labels_csv = 'train_1w_corner.csv'
    val_labels_csv = 'train_b_corner.csv'

    train_dataset.parse_csv(images_dir=train_dir, labels_filename=train_labels_csv,
                            input_format=['image_name', 'class_id', 'xmin', 'ymin', 'xmax', 'ymax'])
    val_dataset.parse_csv(images_dir=val_dir, labels_filename=val_labels_csv,
                            input_format=['image_name', 'class_id', 'xmin', 'ymin', 'xmax', 'ymax'])

    # 3: Set the batch size.

    batch_size = 16 # Change the batch size if you like, or if you run into GPU memory issues.

    # 4: Set the image transformations for pre-processing and data augmentation options.
    # For the training generator: not complete
    ssd_data_augmentation = SSDDataAugmentation(img_height=img_height,
                                                img_width=img_width,
                                                background=mean_color)

    # For the validation generator:
    convert_to_3_channels = ConvertTo3Channels()
    resize = Resize(height=img_height, width=img_width)

    # 5: Instantiate an encoder that can encode ground truth labels into the format needed by the SSD loss function.

    # The encoder constructor needs the spatial dimensions of the model's predictor layers to create the anchor boxes.
    predictor_sizes = [model.get_layer('conv4_3_norm_mbox_conf').output_shape[1:3],
                       model.get_layer('fc7_mbox_conf').output_shape[1:3],
                       model.get_layer('conv6_2_mbox_conf').output_shape[1:3],
                       model.get_layer('conv7_2_mbox_conf').output_shape[1:3],
                       model.get_layer('conv8_2_mbox_conf').output_shape[1:3],
                       model.get_layer('conv9_2_mbox_conf').output_shape[1:3],
                       model.get_layer('conv10_2_mbox_conf').output_shape[1:3]]

    ssd_input_encoder = SSDInputEncoder(img_height=img_height,
                                        img_width=img_width,
                                        n_classes=n_classes,
                                        predictor_sizes=predictor_sizes,
                                        scales=scales,
                                        aspect_ratios_per_layer=aspect_ratios,
                                        two_boxes_for_ar1=two_boxes_for_ar1,
                                        steps=steps,
                                        offsets=offsets,
                                        clip_boxes=clip_boxes,
                                        variances=variances,
                                        matching_type='multi',
                                        pos_iou_threshold=0.5,
                                        neg_iou_limit=0.5,
                                        normalize_coords=normalize_coords)

    # 6: Create the generator handles that will be passed to Keras' `fit_generator()` function.

    train_generator = train_dataset.generate(batch_size=batch_size,
                                             shuffle=True,
                                             transformations=[ssd_data_augmentation],
                                             label_encoder=ssd_input_encoder,
                                             returns={'processed_images',
                                                      'encoded_labels'},
                                             keep_images_without_gt=False)

    val_generator = val_dataset.generate(batch_size=batch_size,
                                         shuffle=False,
                                         transformations=[convert_to_3_channels,
                                                          resize],
                                         label_encoder=ssd_input_encoder,
                                         returns={'processed_images',
                                                  'encoded_labels'},
                                         keep_images_without_gt=False)

    # Get the number of samples in the training and validations datasets.
    train_dataset_size = train_dataset.get_dataset_size()
    val_dataset_size   = val_dataset.get_dataset_size()

    print("Number of images in the training dataset:\t{:>6}".format(train_dataset_size))
    print("Number of images in the validation dataset:\t{:>6}".format(val_dataset_size))

    # Define a learning rate schedule.

    def lr_schedule(epoch):
        if epoch < 80:
            return 0.001
        elif epoch < 100:
            return 0.0001
        else:
            return 0.00001

    # Define model callbacks.

    # TODO: Set the filepath under which you want to save the model.
    model_checkpoint = ModelCheckpoint(filepath='ssd512_1epoch-{epoch:02d}_loss-{loss:.4f}_val_loss-{val_loss:.4f}.h5',
                                       monitor='val_loss',
                                       verbose=1,
                                       save_best_only=False,
                                       save_weights_only=False,
                                       mode='auto',
                                       period=1)

    csv_logger = CSVLogger(filename='ssd512_training_log_1.csv',
                           separator=',',
                           append=True)

    learning_rate_scheduler = LearningRateScheduler(schedule=lr_schedule,
                                                    verbose=1)

    terminate_on_nan = TerminateOnNaN()

    callbacks = [model_checkpoint,
                 csv_logger,
                 learning_rate_scheduler,
                 terminate_on_nan]

    # If you're resuming a previous training, set `initial_epoch` and `final_epoch` accordingly.
    initial_epoch = 0
    final_epoch = 120
    steps_per_epoch = 1000

    history = model.fit_generator(generator=train_generator,
                                  steps_per_epoch=steps_per_epoch,
                                  epochs=final_epoch,
                                  callbacks=callbacks,
                                  validation_data=val_generator,
                                  validation_steps=ceil(val_dataset_size/batch_size),
                                  initial_epoch=initial_epoch)

if __name__ == '__main__':
    # starting...............
    img_height = 512 # Height of the model input images
    img_width = 512 # Width of the model input images
    img_channels = 3 # Number of color channels of the model input images
    mean_color = [123, 117, 104] # The per-channel mean of the images in the dataset. Do not change this value if you're using any of the pre-trained weights.
    swap_channels = [2, 1, 0] # The color channel order in the original SSD is BGR, so we'll have the model reverse the color channel order of the input images.
    n_classes = 1 # Number of positive classes, e.g. 20 for Pascal VOC, 80 for MS COCO
    scales_pascal = [0.07, 0.15, 0.3, 0.45, 0.6, 0.75, 0.9, 1.05] # The anchor box scaling factors used in the original SSD300 for the Pascal VOC datasets
    scales_coco = [0.04, 0.1, 0.26, 0.42, 0.58, 0.74, 0.9, 1.06] # The anchor box scaling factors used in the original SSD300 for the MS COCO datasets
    scales = scales_pascal
    aspect_ratios = [[1.0, 2.0, 0.5],
                     [1.0, 2.0, 0.5, 3.0, 1.0/3.0],
                     [1.0, 2.0, 0.5, 3.0, 1.0/3.0],
                     [1.0, 2.0, 0.5, 3.0, 1.0/3.0],
                     [1.0, 2.0, 0.5, 3.0, 1.0/3.0],
                     [1.0, 2.0, 0.5],
                     [1.0, 2.0, 0.5]] # The anchor box aspect ratios used in the original SSD300; the order matters
    two_boxes_for_ar1 = True
    steps = [8, 16, 32, 64, 128, 256, 512] # The space between two adjacent anchor box center points for each predictor layer.
    offsets = [0.5, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5] # The offsets of the first anchor box center points from the top and left borders of the image as a fraction of the step size for each predictor layer.
    clip_boxes = False # Whether or not to clip the anchor boxes to lie entirely within the image boundaries
    variances = [0.1, 0.1, 0.2, 0.2] # The variances by which the encoded target coordinates are divided as in the original implementation
    normalize_coords = True

    train_main()

```

**Comment by pierluigiferrari** [comment]
I would assume that either the ground truth box coordinates are being encoded incorrectly, or the raw model output is being decoded incorrectly, or both. A few points to debug this:

1. What does the code for decoding the output and drawing the predicted boxes onto the images look like?
2. Do the following: Generate a batch with encoded labels, then decode the encoded labels and print them onto the respective image. Do the ground truth boxes come out correctly? If not, please provide the code that performs the aforementioned procedure.

**Comment by kemingzeng** [comment]
After trying several times, I just found it become absolutely normal....
Still thanks for your patience.
